### PR TITLE
ENH: Add annotations for five array construction functions

### PR DESCRIPTION
**Pull request by BvB93**
This pull request adds annotations to the five array construction functions in `np.core._asarray`:

* `asarray`
* `asanyarray`
* `ascontiguousarray`
* `asfortranarray`
* `require`

### CODE REVIEW

File: numpy/core/_asarray.pyi [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+def asarray(
+    a: object,
+    dtype: DtypeLike = ...,
+    order: _OrderKACF = ...,
+    *,
+    like: ArrayLike = ...
+) -> ndarray: ...
+@overload
+def asanyarray(
+    a: _ArrayType,
```
Comment (BvB93) [inline numpy/core/_asarray.pyi:23]: `def(a: _ArrayType, ...) -> _ArrayType` implies that both the array (sub-)class, dtype and shape are identical for the input and output.

Without higher-kindred typevar we, unfortunately, can not describe a situation were an `ndarray` subclass is returned with a dtype that is different from the initial input (hence the `dtype: None` in this overload).

File: numpy/core/_asarray.pyi [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    dtype: None = ...,
+    requirements: Union[None, _Requirements, Iterable[_Requirements]] = ...,
+    *,
+    like: ArrayLike = ...
+) -> _ArrayType: ...
+@overload
+def require(
+    a: object,
+    dtype: DtypeLike = ...,
+    requirements: Union[_E, Iterable[Union[_E, _Requirements]]] = ...,
```
Comment (BvB93) [inline numpy/core/_asarray.pyi:65]: `Iterable[Union[_E, _Requirements]]`: an iterable which contains an element from `E` and may or may not contain additional elements from  `_Requirements`.

File: numpy/core/_asarray.pyi [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+) -> ndarray: ...
+
+_Requirements = Literal[
+    "C", "C_CONTIGUOUS", "CONTIGUOUS",
+    "F", "F_CONTIGUOUS", "FORTRAN",
+    "A", "ALIGNED",
+    "W", "WRITEABLE",
+    "O", "OWNDATA"
+]
+_E = Literal["E", "ENSUREARRAY"]
```
Comment (eric-wieser) [inline numpy/core/_asarray.pyi:51]: Perhaps
```suggestion
_E = Literal["E", "ENSUREARRAY"]
_Requirements_With_E = Union[_Requirements, _E]
```

Comment (BvB93) [reply numpy/core/_asarray.pyi:51]: Good idea; done in https://github.com/numpy/numpy/pull/17326/commits/07476cfb144baf9f6a889076d02b6e80261e3c76.

**Comment by BvB93** [comment]
Rebased to get rid of a merge conflict.

**Comment by BvB93** [comment]
https://github.com/numpy/numpy/pull/17326/commits/71def2cf787a592c72484ae676fac353b14432d7 implements an alternative approach for ensuring that externally-defined stubs are 
picked up in the main `__init__.pyi` file. Using `from x import y as y` is more compact than 
specifying them in `__all__` and, more importantly, it is less prone to merge conflicts.

**Comment by BvB93** [comment]
https://github.com/numpy/numpy/pull/17326/commits/be77442c1facae452d3742415161c4f8203f4a86 removes the placeholder annotations (see https://github.com/numpy/numpy/pull/17406) for the newly annotated functions introduced in this PR. 

Secondly, if there are no further comments then I feel this can get merged.

**Comment by charris** [comment]
Thanks @BvB93 .

**Pull request merged**